drivers/dht/dht.c

```diff
@@ -19,201 +20,260 @@
- *
- * @retval  0       Success
- * @retval  -1      Timeout occurred before level was reached
- */
-static inline int _wait_for_level(gpio_t pin, bool expect, unsigned timeout)
+static void _bit_parse(struct dht_data *arg)
 {
-    while (((gpio_read(pin) > 0) != expect) && timeout) {
-        xtimer_usleep(1);
-        timeout--;
+    int8_t pos = arg->bit_pos++;
+    if (arg->bit) {
+        arg->data[pos / 8] |= (0x80U >> (pos % 8));
     }
+}
 
-    return (timeout > 0) ? 0 : -1;
+static void _busy_wait_read(struct dht_data *arg)
+{
+    uint32_t start = ztimer_now(ZTIMER_USEC);
+    while (arg->bit_pos != 40) {
+        _wait_for_level(arg->pin, 1, start);
+        start = ztimer_now(ZTIMER_USEC);
+        _wait_for_level(arg->pin, 0, start);
+        arg->bit = (ztimer_now(ZTIMER_USEC) - start > READ_THRESHOLD) ? 1 : 0;
+        _bit_parse(arg);
+    }
 }
 
-static int _read(uint8_t *dest, gpio_t pin)
+static int _validate_checksum(uint8_t *data)
 {
-    DEBUG("[dht] read\n");
-    uint16_t res = 0;
-
-    for (int i = 0; i < 8; i++) {
-        uint32_t start, end;
-        res <<= 1;
-        /* measure the length between the next rising and falling flanks (the
-         * time the pin is high - smoke up :-) */
-        if (_wait_for_level(pin, 1, TIMEOUT)) {
-            return -1;
-        }
-        start = xtimer_now_usec();
+    uint8_t sum = 0;
 
-        if (_wait_for_level(pin, 0, TIMEOUT)) {
-            return -1;
-        }
-        end = xtimer_now_usec();
+    for (uint_fast8_t i = 0; i < 4; i++) {
+        sum += data[i];
+    }
+    if (sum != data[BYTEPOS_CHECKSUM]) {
+        return -EIO;
+    }
+    return 0;
+}
 
-        /* if the high phase was more than 40us, we got a 1 */
-        if ((end - start) > PULSE_WIDTH_THRESHOLD) {
-            res |= 0x0001;
+static int _parse_raw_values(dht_t *dev, uint8_t *data)
+{
+    bool is_negative;
+
+    switch (dev->params.type) {
+    case DHT11:
+    case DHT11_2022:
+        DEBUG_PUTS("[dht] parse raw values with DHT11 data format");
+        dev->last_val.humidity = data[BYTEPOS_HUMIDITY_HIGH] * 10
+                               + data[BYTEPOS_HUMIDITY_LOW];
+        /* MSB for integral temperature byte gives sign, remaining is
+         * abs() of value (beware: this is not two's complement!) */
+        is_negative = data[BYTEPOS_TEMPERATURE_LOW] & 0x80;
+        data[BYTEPOS_TEMPERATURE_LOW] &= ~0x80;
+        /* 2022-12 aosong.com data sheet uses interprets low bits as
+         * 0.01°C per LSB */
+        if (dev->params.type == DHT11_2022) {
+            data[BYTEPOS_TEMPERATURE_LOW] /= 10;
         }
+        if (data[BYTEPOS_TEMPERATURE_LOW] >= 10) {
+            return -ERANGE;
+        }
+        dev->last_val.temperature = data[BYTEPOS_TEMPERATURE_HIGH] * 10
+                                  + data[BYTEPOS_TEMPERATURE_LOW];
+        break;
+    /* AM2301 == DHT21 == DHT22 (same value in enum),
+     * so both are handled here */
```

Comment:
```suggestion
    /* AM2301 == DHT21 == DHT22 (same value in enum),
     * so all are handled here */
```

Again, please squash/amend directly